packages/api/src/middleware/validation/checkProductValidity.js

```diff
@@ -62,18 +82,37 @@ export function checkProductValidity() {
 
     const trx = await transaction.start(ProductModel.knex());
 
-    // Check name uniqueness
     try {
-      const { farm_id } = req.headers;
-      const { product_id } = req.params;
-      let { type, name } = req.body;
-
       if (product_id) {
         const currentRecord = await ProductModel.query(trx).findById(product_id);
+        // Prevent changing type for now, prevents orphan task type products
+        if (type && type != currentRecord.type) {
+          return res.status(400).send('cannot change product type');
+        }
+
         type = type ?? currentRecord.type;
         name = name ?? currentRecord.name;
       }
 
+      if (!name) {
+        return res.status(400).send('new product must have name');
+      }
+
+      if (!type) {
+        return res.status(400).send('new product must have type');
+      }
+
+      const nonModifiableAssets = typesOfProducts.filter((a) => a !== taskTypeProductMap[type]);
+
+      if (taskTypeProductMap[type] && !req.body[taskTypeProductMap[type]]) {
+        return res.status(400).send('must have product details');
+      }
+
+      if (nonModifiableAssets.some((asset) => Object.hasOwn(req.body, asset))) {
```

Comment:
Yeah I copied the name to be consistent -- if I change it then I think I should change them all so it is searchable / fixable.

That is a strange error, you are right that since the map goes to null for the other products it does nothing for them -- I can delete that. But what you should be seeing is still a 400 from a different check "Must not have additional properties". So I am not quite sure what happened there.

![Screenshot 2024-07-10 at 10 56 45 AM](https://github.com/LiteFarmOrg/LiteFarm/assets/30075973/5c6ea756-b778-4e6b-acdf-8e3ae72b625f)

Reply:
Is it because I was doing a PATCH and you are doing a POST? Sorry I didn't include the whole postman UI just to save space but it was a PATCH to that `product_id` above!

Reply:
Okay I think I found that at least for this case it should have had remove additional properties in the baseController. It should now succeed with extra properties that are not part of nonModifiableAssets.